File: accounts.py
import json
import os
import shutil
from copy import deepcopy
from datetime import datetime

_BASE_DIR = os.environ.get('APP_BASE_DIR', os.path.dirname(os.path.abspath(__file__)))
ACCOUNTS_PATH = os.path.join(_BASE_DIR, 'accounts.json')
BASE_PROFILE_DIR = os.path.join(_BASE_DIR, 'browser-profiles', 'default')

DEFAULT_ACCOUNTS = [
    {'name': 'default', 'profileDir': BASE_PROFILE_DIR, 'label': '主账号', 'status': 'ready', 'lastLogin': None, 'createdAt': None},
]


def loadAccounts():
    if not os.path.exists(ACCOUNTS_PATH):
        return deepcopy(DEFAULT_ACCOUNTS)
    try:
        with open(ACCOUNTS_PATH, 'r', encoding='utf-8') as f:
            accounts = json.load(f)
            return accounts or []
    except (json.JSONDecodeError, OSError):
        bak = ACCOUNTS_PATH + '.bak'
        if os.path.exists(bak):
            try:
                with open(bak, 'r', encoding='utf-8') as f:
                    restored = json.load(f)
                if restored:
                    saveAccounts(restored)
                    return restored
            except (json.JSONDecodeError, OSError):
                pass
        return deepcopy(DEFAULT_ACCOUNTS)


def saveAccounts(accounts):
    if os.path.exists(ACCOUNTS_PATH):
        try:
            shutil.copy2(ACCOUNTS_PATH, ACCOUNTS_PATH + '.bak')
        except OSError:
            pass
    with open(ACCOUNTS_PATH, 'w', encoding='utf-8') as f:
        json.dump(accounts, f, indent=2, ensure_ascii=False)


def getAccount(name):
    accounts = loadAccounts()
    for a in accounts:
        if a['name'] == name:
            return a
    return None


def createAccount(name, label):
    accounts = loadAccounts()
    for a in accounts:
        if a['name'] == name:
            raise Exception('Account "' + name + '" already exists')
    profileDir = os.path.join(_BASE_DIR, 'browser-profiles', name)
    if not os.path.exists(profileDir):
        os.makedirs(profileDir, exist_ok=True)
    account = {
        'name': name,
        'profileDir': profileDir,
        'label': label or name,
        'status': 'needs-login',
        'lastLogin': None,
        'createdAt': datetime.now().isoformat(),
    }
    accounts.append(account)
    saveAccounts(accounts)
    return account


def deleteAccount(name):
    accounts = loadAccounts()
    idx = None
    for i, a in enumerate(accounts):
        if a['name'] == name:
            idx = i
            break
    if idx is None:
        raise Exception('Account "' + name + '" not found')
    account = accounts.pop(idx)
    saveAccounts(accounts)
    if os.path.exists(account['profileDir']):
        shutil.rmtree(account['profileDir'], ignore_errors=True)
    return account


def updateAccount(name, updates):
    accounts = loadAccounts()
    for a in accounts:
        if a['name'] == name:
            a.update(updates)
            saveAccounts(accounts)
            return a
    raise Exception('Account "' + name + '" not found')


def updateAccountStatus(name, status):
    accounts = loadAccounts()
    for a in accounts:
        if a['name'] == name:
            a['status'] = status
            if status == 'ready':
                a['lastLogin'] = datetime.now().isoformat()
            saveAccounts(accounts)
            return

# snake_case aliases
load_accounts = loadAccounts
save_accounts = saveAccounts
get_account = getAccount
create_account = createAccount
delete_account = deleteAccount
update_account = updateAccount
update_account_status = updateAccountStatus

File: accounts.js
const fs = require('fs');
const path = require('path');

const ACCOUNTS_PATH = path.join(__dirname, 'accounts.json');
const BASE_PROFILE_DIR = path.join(__dirname, 'browser-profile');

const DEFAULT_ACCOUNTS = [
  { name: 'default', profileDir: BASE_PROFILE_DIR, label: '主账号', status: 'ready', lastLogin: null, createdAt: null },
];

function loadAccounts() {
  if (!fs.existsSync(ACCOUNTS_PATH)) {
    saveAccounts(DEFAULT_ACCOUNTS);
    return JSON.parse(JSON.stringify(DEFAULT_ACCOUNTS));
  }
  try {
    const data = fs.readFileSync(ACCOUNTS_PATH, 'utf-8');
    return JSON.parse(data);
  } catch {
    // Corrupted file — restore from backup or defaults
    const bak = ACCOUNTS_PATH + '.bak';
    if (fs.existsSync(bak)) {
      try {
        const bakData = fs.readFileSync(bak, 'utf-8');
        const restored = JSON.parse(bakData);
        saveAccounts(restored);
        return restored;
      } catch {}
    }
    saveAccounts(DEFAULT_ACCOUNTS);
    return JSON.parse(JSON.stringify(DEFAULT_ACCOUNTS));
  }
}

function saveAccounts(accounts) {
  // Backup before overwriting
  if (fs.existsSync(ACCOUNTS_PATH)) {
    try { fs.copyFileSync(ACCOUNTS_PATH, ACCOUNTS_PATH + '.bak'); } catch {}
  }
  fs.writeFileSync(ACCOUNTS_PATH, JSON.stringify(accounts, null, 2), 'utf-8');
}

function getAccount(name) {
  return loadAccounts().find(a => a.name === name) || null;
}

function createAccount(name, label) {
  const accounts = loadAccounts();
  if (accounts.find(a => a.name === name)) {
    throw new Error(`Account "${name}" already exists`);
  }
  const profileDir = path.join(__dirname, `browser-profile-${name}`);
  if (!fs.existsSync(profileDir)) {
    fs.mkdirSync(profileDir, { recursive: true });
  }
  const account = {
    name,
    profileDir,
    label: label || name,
    status: 'needs-login',
    lastLogin: null,
    createdAt: new Date().toISOString(),
  };
  accounts.push(account);
  saveAccounts(accounts);
  return account;
}

function deleteAccount(name) {
  if (name === 'default') throw new Error('Cannot delete default account');
  const accounts = loadAccounts();
  const idx = accounts.findIndex(a => a.name === name);
  if (idx === -1) throw new Error(`Account "${name}" not found`);
  const [account] = accounts.splice(idx, 1);
  saveAccounts(accounts);
  // Remove profile directory (best effort)
  if (fs.existsSync(account.profileDir)) {
    fs.rmSync(account.profileDir, { recursive: true, force: true });
  }
  return account;
}

function updateAccount(name, updates) {
  const accounts = loadAccounts();
  const account = accounts.find(a => a.name === name);
  if (!account) throw new Error(`Account "${name}" not found`);
  Object.assign(account, updates);
  saveAccounts(accounts);
  return account;
}

function updateAccountStatus(name, status) {
  const accounts = loadAccounts();
  const account = accounts.find(a => a.name === name);
  if (!account) return;
  account.status = status;
  if (status === 'ready') account.lastLogin = new Date().toISOString();
  saveAccounts(accounts);
}

module.exports = { loadAccounts, saveAccounts, getAccount, createAccount, deleteAccount, updateAccount, updateAccountStatus };
